Sources/App/Controllers/User/Auth/FacebookController.swift

```diff
@@ -0,0 +1,109 @@
+import HTTP
+import Vapor
+import Fluent
+
+final class FacebookController {
+
+  private let drop: Droplet
+  private let config: Config
+  private let photoController: PhotoConroller
+  private let fb = Social.FB.self
+
+  init(drop: Droplet) {
+    self.drop = drop
+    self.config = drop.config
+    self.photoController = PhotoConroller(drop: self.drop)
+  }
+
+  func createOrUpdateUserProfile(use authorizateCode: String) throws -> User {
+
+    let accessToken = try authorizate(by: authorizateCode)
+    let (profile, socialUserId) = try getUserProfile(use: accessToken)
+
+    if let user = try SocialAccount.find(by: socialUserId) {
+      user.name = profile.name
+      user.lastname = profile.lastname
+      user.email = profile.email
+      user.photo = try photoController.downloadAndSavePhoto(for: user, by: profile.photo)
+      try user.save()
+      return user
+    }
+
+    try profile.save()
+
+    if let url = profile.photo, !url.isEmpty {
+      profile.photo = try photoController.downloadAndSavePhoto(for: profile, by: profile.photo)
+      try profile.save()
+    }
+
+    try SocialHelper.storeSocialProfile(for: profile, socialUserId: socialUserId, social: Social.Nets.fb)
+    
+    return profile
+  }
+
+  fileprivate func authorizate(by authorizateCode: String) throws -> String {
```

Comment:
may be `by authorizationCode: String`

Reply:
sure! fixed

Reply:
And rename function to `authorize` here and above.